docs/zkapps/tutorials/04-in-browser-react-ui.mdx

```diff
@@ -0,0 +1,407 @@
+
+Next, we check if there is already a private key in local storage - and if not, we save one there. We also check if the account exists on chain:
+
+```ts
+...
+ 40         await zkappWorkerClient.setActiveInstanceToBerkeley();
+ 41
+ 42         if (localStorage.privateKey == null) {
+ 43           localStorage.privateKey = PrivateKey.random().toBase58();
+ 44         }
+ 45
+ 46         let privateKey = PrivateKey.fromBase58(localStorage.privateKey);
+ 47         let publicKey = privateKey.toPublicKey();
+ 48
+ 49         console.log('using key', publicKey.toBase58());
+ 50
+ 51         console.log('checking if account exists...');
+ 52         const res = await zkappWorkerClient.fetchAccount({ publicKey: publicKey! });
+ 53         const accountExists = res.error == null;
+ 54
+ 55         // TODO
+ 56       }
+...
+```
+
+Continuing, we load our zkApp in the web worker. We load the contract, compile it, create a instance of it at a fixed address, and get its current state:
+
+```ts
+...
+ 53         const accountExists = res.error == null;
+ 54
+ 55         await zkappWorkerClient.loadContract();
+ 56
+ 57         console.log('compiling zkApp');
+ 58         await zkappWorkerClient.compileContract();
+ 59         console.log('zkApp compiled');
+ 60
+ 61         const zkappPublicKey = PublicKey.fromBase58('B62qrBBEARoG78KLD1bmYZeEirUfpNXoMPYQboTwqmGLtfqAGLXdWpU');
+ 62
+ 63         await zkappWorkerClient.initZkappInstance(zkappPublicKey);
+ 64
+ 65         console.log('getting zkApp state...');
+ 66         await zkappWorkerClient.fetchAccount({ publicKey: zkappPublicKey })
+ 67         const currentNum = await zkappWorkerClient.getNum();
+ 68         console.log('current state:', currentNum.toString());
+ 69
+ 70         // TODO
+ 71       }
+...
+```
+
+And lastly for this function, we update the state of the react app:
+
+```ts
+...
+ 68         console.log('current state:', currentNum.toString());
+ 69
+ 70         setState({
+ 71             ...state,
+ 72             zkappWorkerClient,
+ 73             hasBeenSetup: true,
+ 74             publicKey,
+ 75             privateKey,
+ 76             zkappPublicKey,
+ 77             accountExists,
+ 78             currentNum
+ 79         });
+ 80       }
+ 81     })();
+...
+```
+
+Now that we have finished setting up our UI, we write a new effect, that waits for the account to exist, if it didn't exist before. 
+
+If the account has been newly created for example, it will need to be funded from the faucet. We'll add in our UI later a link to request funds for users to request funds while that account does not exist.
+
+```ts
+...
+ 84   // -------------------------------------------------------
+ 85   // Wait for account to exist, if it didn't
+ 86
+ 87   useEffect(() => {
+ 88     (async () => {
+ 89       if (state.hasBeenSetup && !state.accountExists) {
+ 90         for (;;) {
+ 91           console.log('checking if account exists...');
```

Comment:
this feels like a missing snarkyjs feature! get notified on creation of an on-chain account -- so that we don't have to make everyone write these polling loops.

created an issue to discuss for our eng team: https://github.com/o1-labs/snarkyjs/issues/531